src/lua/tweaks.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2022-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "lua/tweaks.h"
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "core/tweaks.h"
+#include "diag.h"
+#include "lua/utils.h"
+#include "trivia/util.h"
+
+/**
+ * Pushes a tweak value to Lua stack and returns the number of values pushed.
+ */
+static int
+luaT_push_tweak_value(struct lua_State *L, const struct tweak_value *v)
+{
+	switch (v->type) {
+	case TWEAK_VALUE_BOOL:
+		lua_pushboolean(L, v->bval);
+		return 1;
+	case TWEAK_VALUE_INT:
+		lua_pushinteger(L, v->ival);
+		return 1;
+	case TWEAK_VALUE_STR:
+		lua_pushstring(L, v->sval);
+		return 1;
+	default:
+		unreachable();
+	}
+}
+
+/**
+ * Looks up a tweak value by name (2nd arg) and pushes it to Lua stack.
+ * Returns nothing if there's no such tweak.
+ */
+static int
+luaT_tweaks_index(struct lua_State *L)
+{
+	const char *name = luaL_checkstring(L, 2);
+	struct tweak *t = tweak_find(name);
+	if (t == NULL)
+		return 0;
+	struct tweak_value v;
+	tweak_get(t, &v);
+	return luaT_push_tweak_value(L, &v);
+}
+
+/**
+ * Updates a tweak value given its name (2nd arg) and the new value (3rd arg).
+ * Raises an error if there's no such tweak or the value is invalid.
+ */
+static int
+luaT_tweaks_newindex(struct lua_State *L)
+{
+	const char *name = luaL_checkstring(L, 2);
+	struct tweak *t = tweak_find(name);
+	if (t == NULL) {
+		diag_set(IllegalParams, "No such option");
+		return luaT_error(L);
+	}
+	struct tweak_value v;
+	switch (lua_type(L, 3)) {
+	case LUA_TBOOLEAN:
+		v.type = TWEAK_VALUE_BOOL;
+		v.bval = lua_toboolean(L, 3);
+		break;
+	case LUA_TNUMBER:
+		v.type = TWEAK_VALUE_INT;
+		v.ival = lua_tointeger(L, 3);
+		if (lua_tonumber(L, 3) != v.ival)
+			goto invalid_value;
+		break;
+	case LUA_TSTRING:
+		v.type = TWEAK_VALUE_STR;
+		v.sval = lua_tostring(L, 3);
+		break;
+	default:
+invalid_value:
+		diag_set(IllegalParams,
+			 "Value must be boolean, integer, or string");
+		return luaT_error(L);
+	}
+	if (tweak_set(t, &v) != 0)
+		return luaT_error(L);
+	return 0;
+}
+
+static bool
+luaT_tweaks_serialize_foreach_cb(const char *name, struct tweak *t, void *arg)
+{
+	struct lua_State *L = arg;
+	struct tweak_value v;
+	tweak_get(t, &v);
+	luaT_push_tweak_value(L, &v);
+	lua_setfield(L, -2, name);
+	return true;
+}
+
+/**
+ * Pushes a table that maps tweak names to their values to Lua stack.
+ */
+static int
+luaT_tweaks_serialize(struct lua_State *L)
+{
+	lua_newtable(L);
+	tweak_foreach(luaT_tweaks_serialize_foreach_cb, L);
+	return 1;
+}
+
+void
+tarantool_lua_tweaks_init(struct lua_State *L)
+{
+	const struct luaL_Reg mt[] = {
+		{"__index", luaT_tweaks_index},
+		{"__newindex", luaT_tweaks_newindex},
+		{"__serialize", luaT_tweaks_serialize},
+		{"__autocomplete", luaT_tweaks_serialize},
+		{NULL, NULL},
+	};
+	luaL_register(L, "internal.tweaks", NULL);
```

Comment:
What do you think about https://github.com/tarantool/doc/issues/3308?

Maybe replace `foo.lib` in my PR (see [here](https://github.com/tarantool/tarantool/pull/8173#discussion_r1088439359)) with `internal.foo` (and use some temporary name for the Lua/C part of compat)? If we'll agree on it, I'll replace `net.box.lib` with `internal.net.box` as well.

Reply:
> Maybe replace `foo.lib` in my PR (see [here](https://github.com/tarantool/tarantool/pull/8173#discussion_r1088439359)) with `internal.foo` (and use some temporary name for the Lua/C part of compat)? If we'll agree on it, I'll replace `net.box.lib` with `internal.net.box` as well.

There's a difference between `net.box.lib` and `internal.tweaks`. The former is a helper module that is supposed to be used by the net.box implementation only while the latter is a final module that can be used by end users. We just don't want the tweaks module to be public. So no, I don't think we should rename `foo.lib` to `internal.foo` in your PR. I think of `require('tarantool')._internal.tweaks` as an alias for `require('internal.tweaks')`. I can't get rid of `internal.tweaks` completely (e.g. by fixing `package.loaded`), because it's used by the `compat` module which is loaded before the `tarantool` module. Maybe, we should drop the `require('tarantool')._internal.tweaks` alias altogether and use `internal.tweaks` instead? Come to think of it, maybe name the new module simply `tweaks`? It sounds "internal" enough.

As for the Lua/C part of compat, I drop it in the last path of the series, because it's not needed anymore.

Reply:
Regarding `foo.lib` and `internal.foo`: okay, let's assume this is our new naming rule.

Regarding eliminating of the `internal.tweak` module: I don't think that it worth to do at all. We need some mechanism to pass values over built-in modules and `require()` machinery fits good here. It makes our built-in modules similar to usual ones and reduces amount of practices a developer should be aware of. I wouldn't change it. At least I don't see a reason for now.

Regarding entrypoint to the `tweaks` API: I don't mind any of the options, but I would prefer to have 'internal' word somewhere or start the name from the underscore symbol.

Reply:
Okay, let's leave `internal.tweaks` for the module name and drop the `tarantool._internal.tweaks` alias, as it isn't used anywhere anyway and users can just as well call `require('internal.tweaks')` instead.

<details><summary>diff</summary>

```diff
diff --git a/src/lua/init.lua b/src/lua/init.lua
index 226afb83c8e7..c8f5e1eef915 100644
--- a/src/lua/init.lua
+++ b/src/lua/init.lua
@@ -356,6 +356,5 @@ return {
         strip_cwd_from_path = strip_cwd_from_path,
         module_name_from_filename = module_name_from_filename,
         run_preload = run_preload,
-        tweaks = require('internal.tweaks'),
     };
 }
```

</details>